dockerfiles/Dockerfile.training

```diff
@@ -0,0 +1,193 @@
+
+# install mkl, numpy, and pytorch build dependencies
+ARG MKL_VERSION
+ARG NUMPY_VERSION
+ARG ONNX_VERSION
+RUN conda install -y \
+        mkl=${MKL_VERSION} \
+        mkl-include=${MKL_VERSION} \
+        numpy=${NUMPY_VERSION} \
+        cmake \
+        ninja \
+        pyyaml \
+        cffi \
+        setuptools &&\
+    pip install \
+        onnx=="${ONNX_VERSION}"
+
+# install mellanox libraries for ib etc.
+ARG MOFED_VERSION
+ARG MOFED_OS=ubuntu18.04
+ARG MOFED_TARNAME=MLNX_OFED_LINUX-${MOFED_VERSION}-${MOFED_OS}-x86_64
+ARG MOFED_URL=https://www.mellanox.com/downloads/ofed/MLNX_OFED-${MOFED_VERSION}/${MOFED_TARNAME}.tgz
+RUN cd /stage && curl -fSsL ${MOFED_URL} | tar -zxpf - &&\
+    mv ${MOFED_TARNAME} mofed-${MOFED_VERSION} &&\
+    cd mofed-${MOFED_VERSION} && \
+    ./mlnxofedinstall --basic --force --user-space-only --without-fw-update
+
+# build ucx suite (leverages mellanox libraries)
+# note: openmpi will not select ucx without multithreading enabled
+ARG UCX_VERSION
+ARG UCX_TARNAME=ucx-$UCX_VERSION
+ARG UCX_URL=https://github.com/openucx/ucx/releases/download/v${UCX_VERSION}/${UCX_TARNAME}.tar.gz
+RUN apt-get -y install \
+        libnuma-dev &&\
+    cd /stage && curl -fSsL ${UCX_URL} | tar xzf - &&\
+    cd ${UCX_TARNAME} &&\
+    ./configure \
+	--prefix=/opt/ucx \
+        --with-cuda=/usr/local/cuda \
+        --with-verbs=/usr \
+        --enable-mt &&\
+    make -j"$(nproc)" &&\
+    make install
+
+# build openmpi (use --prefix /opt/ompi4 to move to runtime image)
+# note: require --enable-orterun-prefix-by-default for Azure machine learning compute
+# note: disable verbs as we use ucx middleware and don't want btl openib warnings
+ARG OPENMPI_VERSION
+ARG OPENMPI_TARNAME=openmpi-${OPENMPI_VERSION}
+ARG OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v%OMPI_BASE%/${OPENMPI_TARNAME}.tar.gz
+RUN export OMPI_BASE=${OPENMPI_VERSION%.*} &&\
+    cd /stage && curl -fSsL `echo ${OPENMPI_URL} | sed s/%OMPI_BASE%/$OMPI_BASE/` | tar xzf - &&\
+    cd ${OPENMPI_TARNAME} &&\
+    ./configure \
+        --prefix=/opt/ompi4 \
+        --with-ucx=/opt/ucx \
+        --without-verbs \
+        --with-cuda=/usr/local/cuda \
+        --enable-mpirun-prefix-by-default \
+        --enable-orterun-prefix-by-default \
+        --enable-mca-no-build=btl-uct &&\
+    make -j"$(nproc)" install &&\
+    ldconfig
+ENV PATH=/opt/ompi4/bin:$PATH
+ENV LD_LIBRARY_PATH=/opt/ompi4/lib:$LD_LIBRARY_PATH
+
+# install mpi4py (be sure to link existing /opt/ompi4)
+RUN CC=mpicc MPICC=mpicc pip install mpi4py --no-binary mpi4py
+
+# build pytorch (using v1.6.0-rc3 + patch)
+# todo: need to update to pytorch 1.6.0 releaase when ready
+# note: FindNCCL.patch is in pytorch master
+COPY FindNCCL.patch /stage/
+ENV CUDA_HOME="/usr/local/cuda" \
+    CUDNN_LIBRARY="/usr/lib/x86_64-linux-gnu" \
+    NCCL_INCLUDE_DIR="/usr/local/cuda/include" \
+    NCCL_LIB_DIR="/usr/local/cuda/lib64" \
+    USE_SYSTEM_NCCL=1 
+RUN apt-get -y install git &&\
```

Comment:
```suggestion
RUN apt update && apt install -y git &&\
```

Reply:
So is there a reason to prefer apt over apt-get? I read online that apt is targeted for console feedback while apt-get is targeted for scripts and logs (this [StackExchange](https://askubuntu.com/questions/990823/apt-gives-unstable-cli-interface-warning)).

Reply:
`apt-get` is preferred only if you plan to **parse** the output of commands. Otherwise, `apt` is a newer tool with more features than `apt-get` (`apt` accumulates functionalities from several tools, like `apt-get`, `apt-cache`, `dpkg list`, etc) and has a more verbose output which is nice for debug.

I prefer `apt` as it is shorter to type and provide all features, but it is just a preference.

Reply:
I only found references to apt-get on Docker website and Docker images elsewhere. And apt prints messages about unstable cli interface. So I think best to stick with apt-get for now.